File: library/src/main/java/me/xiaopan/android/spear/cache/LruMemoryCache.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@xiaopan.me>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.android.spear.cache;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.text.format.Formatter;
import android.util.Log;

import me.xiaopan.android.spear.RecycleDrawable;
import me.xiaopan.android.spear.Spear;
import me.xiaopan.android.spear.util.LruCache;

/**
 * 使用Lru算法来缓存位图
 */
public class LruMemoryCache implements MemoryCache {
	private static final String NAME = "LruMemoryCache";
	private Context context;
	private LruCache<String, Drawable> drawableLruCache;

	public LruMemoryCache(Context context, int maxSize){
		this.context = context;
        this.drawableLruCache = new DrawableLruCache(maxSize);
		if(Spear.isDebugMode()){
			Log.i(Spear.TAG, NAME + " - " + "MemoryCacheMaxSize: "+ Formatter.formatFileSize(context, maxSize));
		}
	}

    public LruMemoryCache(Context context){
		this(context, (int) (Runtime.getRuntime().maxMemory()/8));
	}
	
	@Override
	public synchronized void put(String key, Drawable value) {
		if(!(value instanceof RecycleDrawable)){
			throw new IllegalArgumentException("drawable must be implemented RecycleDrawable");
		}
		drawableLruCache.put(key, value);
		if(Spear.isDebugMode()){
			Log.i(Spear.TAG, NAME + " - " + "put" + " - " + "MemoryCacheSize: "+ Formatter.formatFileSize(context, drawableLruCache.size()));
		}
	}

	@Override
	public synchronized Drawable get(String key) {
		return drawableLruCache.get(key);
	}

	@Override
	public synchronized Drawable remove(String key) {
		Drawable drawable = drawableLruCache.remove(key);
		if(Spear.isDebugMode()){
			Log.i(Spear.TAG, NAME + " - " + "remove" + " - "  + "MemoryCacheSize: "+ Formatter.formatFileSize(context, drawableLruCache.size()));
		}
		return drawable;
	}

	@Override
	public long getSize() {
		return drawableLruCache.size();
	}

	@Override
	public long getMaxSize() {
		return drawableLruCache.maxSize();
	}

	@Override
	public synchronized void clear() {
		if(Spear.isDebugMode()){
			Log.i(Spear.TAG, NAME + " - " + "clear" + " - "  + "before clean MemoryCacheSize: "+ Formatter.formatFileSize(context, drawableLruCache.size()));
		}
		drawableLruCache.evictAll();
	}

	private class DrawableLruCache extends LruCache<String, Drawable> {

		public DrawableLruCache(int maxSize) {
			super(maxSize);
		}

		@Override
		public Drawable put(String key, Drawable value) {
			((RecycleDrawable) value).setIsCached(NAME+":put", true);
			return super.put(key, value);
		}

		@Override
		protected int sizeOf(String key, Drawable value) {
			int bitmapSize = ((RecycleDrawable) value).getSize();
			return bitmapSize == 0 ? 1 : bitmapSize;
		}

		@Override
		protected void entryRemoved(boolean evicted, String key, Drawable oldValue, Drawable newValue) {
			((RecycleDrawable) oldValue).setIsCached(NAME+":entryRemoved", false);
		}
	}
}

File: sketch/src/main/java/com/github/panpf/sketch/cache/internal/LruMemoryCache.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.cache.internal

import android.content.ComponentCallbacks2
import com.github.panpf.sketch.cache.CountBitmap
import com.github.panpf.sketch.cache.MemoryCache
import com.github.panpf.sketch.util.Logger
import com.github.panpf.sketch.util.LruCache
import com.github.panpf.sketch.util.format
import com.github.panpf.sketch.util.formatFileSize
import com.github.panpf.sketch.util.getTrimLevelName
import com.github.panpf.sketch.util.toHexString
import kotlinx.coroutines.sync.Mutex
import java.util.WeakHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * A bitmap memory cache that manages the cache according to a least-used rule
 */
class LruMemoryCache constructor(override val maxSize: Long) : MemoryCache {

    companion object {
        private const val MODULE = "LruMemoryCache"

        @JvmStatic
        private val editLockLock = Any()
    }

    private val cache: LruCache<String, CountBitmap> = CountBitmapLruCache(maxSize)
    private val editLockMap: MutableMap<String, Mutex> = WeakHashMap()
    private val getCount = AtomicInteger()
    private val hitCount = AtomicInteger()

    override var logger: Logger? = null
    override val size: Long
        get() = cache.size().toLong()

    override fun put(key: String, countBitmap: CountBitmap): Boolean {
        return if (cache[key] == null) {
            cache.put(key, countBitmap)
            logger?.d(MODULE) {
                "put. ${countBitmap.info}. ${size.formatFileSize()}. $key"
            }
            true
        } else {
            logger?.w(MODULE, "Exist. key=$key")
            false
        }
    }

    override fun remove(key: String): CountBitmap? =
        cache.remove(key).apply {
            logger?.d(MODULE) {
                "remove. ${this.info}. ${size.formatFileSize()}. $key"
            }
        }

    override fun get(key: String): CountBitmap? =
        cache[key]?.takeIf {
            (!it.isRecycled).apply {
                if (!this) {
                    cache.remove(key)
                }
            }
        }.apply {
            val getCount1 = getCount.addAndGet(1)
            val hitCount1 = if (this != null) {
                hitCount.addAndGet(1)
            } else {
                hitCount.get()
            }
            if (getCount1 == Int.MAX_VALUE || hitCount1 == Int.MAX_VALUE) {
                getCount.set(0)
                hitCount.set(0)
            }
            logger?.d(MODULE) {
                if (this != null) {
                    val hitRatio = (hitCount1.toFloat() / getCount1).format(2)
                    "get. Hit(${hitRatio}). ${this.info}/${this.bitmap!!.toHexString()}. $key"
                } else {
                    val hitRatio = (hitCount1.toFloat() / getCount1).format(2)
                    "get. NoHit(${hitRatio}). $key"
                }
            }
        }

    override fun exist(key: String): Boolean =
        cache[key]?.takeIf {
            (!it.isRecycled).apply {
                if (!this) {
                    cache.remove(key)
                }
            }
        } != null

    override fun trim(level: Int) {
        val oldSize = size
        if (level >= ComponentCallbacks2.TRIM_MEMORY_MODERATE) {
            cache.evictAll()
        } else if (level >= ComponentCallbacks2.TRIM_MEMORY_BACKGROUND) {
            cache.trimToSize(cache.maxSize() / 2)
        }
        val releasedSize = oldSize - size
        logger?.w(
            MODULE,
            "trim. level '${getTrimLevelName(level)}', released ${releasedSize.formatFileSize()}, size ${size.formatFileSize()}"
        )
    }

    override fun clear() {
        val oldSize = size
        cache.evictAll()
        logger?.w(MODULE, "clear. cleared ${oldSize.formatFileSize()}")
    }

    override fun editLock(key: String): Mutex = synchronized(editLockLock) {
        editLockMap[key] ?: Mutex().apply {
            this@LruMemoryCache.editLockMap[key] = this
        }
    }

    override fun toString(): String = "$MODULE(maxSize=${maxSize.formatFileSize()})"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as LruMemoryCache

        if (maxSize != other.maxSize) return false

        return true
    }

    override fun hashCode(): Int {
        return maxSize.hashCode()
    }


    private class CountBitmapLruCache constructor(maxSize: Long) :
        LruCache<String, CountBitmap>(maxSize.toInt()) {

        override fun put(key: String, countBitmap: CountBitmap): CountBitmap? {
            countBitmap.setIsCached(true, MODULE)
            return super.put(key, countBitmap)
        }

        override fun sizeOf(key: String, countBitmap: CountBitmap): Int {
            val bitmapSize = countBitmap.byteCount
            return if (bitmapSize == 0) 1 else bitmapSize
        }

        override fun entryRemoved(
            evicted: Boolean, key: String, oldCountBitmap: CountBitmap, newCountBitmap: CountBitmap?
        ) {
            oldCountBitmap.setIsCached(false, MODULE)
        }
    }
}